packages/expo-video/android/src/main/java/expo/modules/video/managers/PictureInPictureManager.kt

```diff
@@ -0,0 +1,329 @@
+    get() = videoViews.mapNotNull { it.get() }
+  private val rootViewChildrenOriginalVisibility: MutableMap<Int, Int> = mutableMapOf()
+  private var isReleased = false
+
+  init {
+    VideoManager.registerListener(this)
+  }
+
+  fun release() {
+    if (isInPiP) {
+      // VideoView most likely doesn't exist atp
+      layoutForPiPExit(null)
+    }
+
+    videoViews.clear()
+    removePiPHelperFragment()
+    isReleased = true
+  }
+
+  /**
+   * Updates whether the main activity will auto-enter Pictrue in Picture, elects a new PiP candidate
+   * and refreshes PiP parameters such as the rectHint.
+   */
+  private fun findAndSetupPipCandidate() {
+    val newCandidate = findAutoPiPViewCandidate(strongVideoViews)
+    currentPiPViewCandidate = WeakReference(newCandidate)
+
+    val noBlockingView = strongVideoViews.none { it.pipParams.blocksAppFromEntering }
+    val newAutoEnter = noBlockingView && strongVideoViews.any { it.pipParams.autoEnter && it.pipParams.canEnter }
+    autoEnterPiP = newAutoEnter
+
+    if (!newAutoEnter) {
+      mainActivity.get()?.let {
+        applyPiPParams(it, autoEnterPiP)
+      }
+    }
+
+    newCandidate?.let {
+      applyPipParamsForView(newCandidate)
+    }
+  }
+
+  fun enterPictureInPicture(videoView: VideoView) {
+    val mainActivity = mainActivity.get() ?: return
+
+    currentPiPViewCandidate = WeakReference(videoView)
+    applyPipParamsForView(videoView)
+
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+      mainActivity.enterPictureInPictureMode(PictureInPictureParams.Builder().build())
+    } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+      @Suppress("DEPRECATION")
+      mainActivity.enterPictureInPictureMode()
+    }
+  }
+
+  private fun applyPipParamsForView(view: VideoView) {
+    mainActivity.get()?.let {
+      applyRectHint(it, calculateRectHint(view.playerView))
+      applyPiPParams(it, autoEnterPiP, view.pipParams.aspectRatio)
+    }
+  }
+
+  override fun onPiPParamsChanged(videoView: VideoView, oldPiPParams: PiPParams, newPiPParams: PiPParams) {
+    val pipParamsInfluenceRectHint = pipParamChangeInfluenceRectHint(oldPiPParams, newPiPParams)
+    if (paramChangeInfluencesAutoEnter(oldPiPParams, newPiPParams)) {
+      findAndSetupPipCandidate()
+    } else if (videoView == currentPiPViewCandidate.get() && pipParamsInfluenceRectHint) {
+      applyPipParamsForView(videoView)
+    }
+  }
+
+  private fun pipParamChangeInfluenceRectHint(old: PiPParams, new: PiPParams): Boolean {
+    return old.rectHint != new.rectHint || old.aspectRatio != new.aspectRatio
+  }
+
+  override fun onPictureInPictureStart() {
+    maybeWarnAboutAutoEnterViews()
+
+    val candidate = currentPiPViewCandidate.get()
+    candidate?.let {
+      layoutForPiPEnter(it)
+    }
+
+    strongVideoViews.forEach {
+      it.onStartPictureInPicture(candidate)
+    }
+
+    isInPiP = true
+  }
+
+  override fun onPictureInPictureStop() {
+    if (mainActivity.get() == null) {
```

Comment:
We're checking if we have a reference to the main activity, but should we check the `isInPip` instead?

Reply:
You are right, it's a leftover from development when onPictureInPictureStop contained code from layoutForPiPExit